src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -79,41 +80,93 @@ internal static ValueTask<int> ReadAtOffsetAsync(SafeFileHandle handle, Memory<b
             long fileOffset, CancellationToken cancellationToken)
             => ScheduleSyncReadScatterAtOffsetAsync(handle, buffers, fileOffset, cancellationToken);
 
-        internal static unsafe int WriteAtOffset(SafeFileHandle handle, ReadOnlySpan<byte> buffer, long fileOffset)
+        internal static unsafe void WriteAtOffset(SafeFileHandle handle, ReadOnlySpan<byte> buffer, long fileOffset)
         {
-            fixed (byte* bufPtr = &MemoryMarshal.GetReference(buffer))
+            while (true)
             {
-                // The Windows implementation uses WriteFile, which ignores the offset if the handle
-                // isn't seekable.  We do the same manually with PWrite vs Write, in order to enable
-                // the function to be used by FileStream for all the same situations.
-                int result = handle.CanSeek ?
-                    Interop.Sys.PWrite(handle, bufPtr, buffer.Length, fileOffset) :
-                    Interop.Sys.Write(handle, bufPtr, buffer.Length);
-                FileStreamHelpers.CheckFileCall(result, handle.Path);
-                return result;
+                fixed (byte* bufPtr = &MemoryMarshal.GetReference(buffer))
+                {
+                    // The Windows implementation uses WriteFile, which ignores the offset if the handle
+                    // isn't seekable.  We do the same manually with PWrite vs Write, in order to enable
+                    // the function to be used by FileStream for all the same situations.
+                    int bytesWritten = handle.CanSeek ?
+                        Interop.Sys.PWrite(handle, bufPtr, GetNumberOfBytesToWrite(buffer.Length), fileOffset) :
+                        Interop.Sys.Write(handle, bufPtr, GetNumberOfBytesToWrite(buffer.Length));
+
+                    FileStreamHelpers.CheckFileCall(bytesWritten, handle.Path);
+                    if (bytesWritten == buffer.Length || bytesWritten == 0)
```

Comment:
in what scenario `bytesWritten == 0` but the buffer `buffer.Length != 0`? Is it not an incomplete write that we were supposed to fix?
```suggestion
                    if (bytesWritten == buffer.Length)
```

Reply:
> in what scenario bytesWritten == 0 but the buffer buffer.Length != 0?

I was just being overly cautious, so as to avoid any possibility of an infinite loop.  I'll remove it.